File: solutions/java/src/main/java/com/epi/SpiralMatrixClockwise.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class SpiralMatrixClockwise {
  // @include
  public static List<Integer> matrixInSpiralOrder(
      List<List<Integer>> squareMatrix) {
    List<Integer> spiralOrdering = new ArrayList<>();
    for (int offset = 0; offset < Math.ceil(0.5 * squareMatrix.size());
         ++offset) {
      matrixLayerInClockwise(squareMatrix, offset, spiralOrdering);
    }
    return spiralOrdering;
  }

  private static void matrixLayerInClockwise(List<List<Integer>> squareMatrix,
                                             int offset,
                                             List<Integer> spiralOrdering) {
    if (offset == squareMatrix.size() - offset - 1) {
      // squareMatrix has odd dimension, and we are at its center.
      spiralOrdering.add(squareMatrix.get(offset).get(offset));
      return;
    }

    for (int j = offset; j < squareMatrix.size() - offset - 1; ++j) {
      spiralOrdering.add(squareMatrix.get(offset).get(j));
    }
    for (int i = offset; i < squareMatrix.size() - offset - 1; ++i) {
      spiralOrdering.add(
          squareMatrix.get(i).get(squareMatrix.size() - offset - 1));
    }
    for (int j = squareMatrix.size() - offset - 1; j > offset; --j) {
      spiralOrdering.add(
          squareMatrix.get(squareMatrix.size() - offset - 1).get(j));
    }
    for (int i = squareMatrix.size() - offset - 1; i > offset; --i) {
      spiralOrdering.add(squareMatrix.get(i).get(offset));
    }
  }
  // @exclude

  private static void simpleTest() {
    List<List<Integer>> A = Arrays.asList(
        Arrays.asList(1, 2, 3), Arrays.asList(4, 5, 6), Arrays.asList(7, 8, 9));
    List<Integer> goldenResult = Arrays.asList(1, 2, 3, 6, 9, 8, 7, 4, 5);
    List<Integer> result = matrixInSpiralOrder(A);
    assert(result.equals(goldenResult));
  }

  public static void main(String[] args) {
    simpleTest();
    Random gen = new Random();
    int N;
    if (args.length == 1) {
      N = Integer.parseInt(args[0]);
    } else {
      N = gen.nextInt(50) + 1;
    }
    List<List<Integer>> A = new ArrayList<>(N);
    int x = 1;
    for (int i = 0; i < N; ++i) {
      A.add(new ArrayList(N));
      for (int j = 0; j < N; ++j) {
        A.get(i).add(x++);
      }
    }
    List<Integer> result = matrixInSpiralOrder(A);
    for (Integer a : result) {
      System.out.print(a + " ");
    }
    System.out.println();
  }
}

File: solutions/cpp/Spiral_matrix_clockwise.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

void PrintMatrixClockwise(const vector<vector<int>> &A, int offset);

// @include
void PrintMatrixInSpiralOrder(const vector<vector<int>> &A) {
  for (int offset = 0; offset < ceil(0.5 * A.size()); ++offset) {
    PrintMatrixClockwise(A, offset);
  }
}

void PrintMatrixClockwise(const vector<vector<int>> &A, int offset) {
  if (offset == A.size() - offset - 1) {
    // A has odd diemnsion, and we are at the center of the matrix A.
    cout << A[offset][offset];
    return;
  }

  for (int j = offset; j < A.size() - offset - 1; ++j) {
    cout << A[offset][j] << ' ';
  }
  for (int i = offset; i < A.size() - offset - 1; ++i) {
    cout << A[i][A.size() - offset - 1] << ' ';
  }
  for (int j = A.size() - offset - 1; j > offset; --j) {
    cout << A[A.size() - offset - 1][j] << ' ';
  }
  for (int i = A.size() - offset - 1; i > offset; --i) {
    cout << A[i][offset] << ' ';
  }
}
// @exclude

int main(int argc, char *argv[]) {
  default_random_engine gen((random_device())());
  int N;
  if (argc == 2) {
    N = atoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 50);
    N = dis(gen);
  }
  vector<vector<int>> A(N, vector<int>(N));
  int x = 1;
  for (size_t i = 0; i < N; ++i) {
    for (size_t j = 0; j < N; ++j) {
      A[i][j] = x++;
    }
  }
  PrintMatrixInSpiralOrder(A);
  return 0;
}
